clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -400,12 +412,33 @@ private void handleFetchResponse(ClientResponse resp, FetchRequest request) {
             int totalBytes = 0;
             int totalCount = 0;
             FetchResponse response = new FetchResponse(resp.responseBody());
+
+            // Look for OffsetOutOfRange error in fetchResponse
+            Map<TopicPartition, Long> offsetOutOfRangePartitions = new HashMap();
+            for (Map.Entry<TopicPartition, FetchResponse.PartitionData> entry : response.responseData().entrySet()) {
```

Comment:
Actually I have thought of this. I agree that, with proper implementation we can use the partition data that doesn't avoid error so that we save bandwidth. But I am not doing it because it is harder to reason the code when your records returned from poll() may contains data from "multiple" fetches -- if you have reassignment you may have reordered data as well.

It is certainly possible to think through all the scenario to make it right. But this complicates both this and future implementation. It doesn't seem worth it because the OffsetOutofRange exception should be pretty rare and the bandwidth used to re-fetch data should be very small.